### ISSUE TITLE
panic location points to start of call chain instead of expect call

### ISSUE THREAD

**Issue by est31** [rust-lang/rust#69977]
In call chains, the newly stabilized feature to make panics point to useful locations doesn't report the actual call site of the expand/unwrap call, but to the start of the call chain.

```Rust
    let v = None;
   //v---- reported panic location
    v.map(|x: ()| x)
        .map(|_| ())
      //v---- expected panic location 
        .expect("test");
```

It should rather point to the expect/unwrap instead. You can have multiple such expects/unwraps in your chain.

Edit: happens on latest nightly as well as on the 1.42.0 stable release.

**Comment by Centril** [comment]
cc @eddyb @anp

**Comment by eddyb** [comment]
This is because of the `Span` of the call is the whole expression.
Diagnostics also run into this, but at least there you can see the whole range, not just the starting point.
We could make `core::panic::Location` keep a range but that would increase its size.

For method calls, it would be ideal to point to the method identifier, but propagating that information through MIR might be non-trivial.

cc @rust-lang/wg-diagnostics @rust-lang/wg-mir-opt

**Comment by anp** [comment]
In the course of working on track_caller I've actually wondered whether `core::panic::Location` would be better implemented internally as containing a span, and then using *handwaving intrinsic magic* to implement the file/line/col information on top of that. A bonus would be the ability to implement `location-detail-control` from the RFC and e.g. eliminate the filename strings without losing unique `Location` pointers.

**Comment by eddyb** [comment]
The "*handwaving intrinsic magic*" would much easier go into the `Location` value itself rather than having some generated side-tables or whatever.
Anyway, we could add an "end" line/column pair, I wonder if that would slow anything down.

**Comment by anp** [comment]
Looks like the fix for this is being tracked in https://github.com/rust-lang/rust/issues/68686, am I correct?
